pkg/urbit/vere/io/ames.c

```diff
@@ -748,188 +909,227 @@ _ames_lane_scry_cb(void* vod_p, u3_noun nun)
+    //
+    else if ( u3_nul == lac ) {
+      c3_free(hun_y);
+      return;
     }
   }
 
-  //  ensure the mug is valid
+  //  proceed with forwarding
   //
-  if ( c3y == pas_o
-    && (hed_u.mug_l != _ca_mug_body(nrd_i - 4, bod_y)) )
   {
-    pas_o = c3n;
+    //  store the packet details for later processing
+    //
+    //    XX allocates unnecessarily when we know the lane
+    //
+    u3_panc* pac_u = c3_calloc(sizeof(*pac_u));
+    pac_u->sam_u = sam_u;
+    pac_u->hed_u = *hed_u;
+    pac_u->bod_u = *bod_u;
+    pac_u->ore_u = *lan_u;
+    pac_u->ptr_v = hun_y;
+
+    if ( 0 != sam_u->pac_u ) {
+      pac_u->nex_u = sam_u->pac_u;
+      sam_u->pac_u->pre_u = pac_u;
+    }
+    sam_u->pac_u = pac_u;
 
-    sam_u->mut_d++;
-    if ( 0 == (sam_u->mut_d % 100) ) {
-      u3l_log("ames: %" PRIu64 " dropped for invalid mug\n", sam_u->mut_d);
+    //  if we already know the lane, just forward
+    //
+    if ( u3_none != lac ) {
+      _ames_forward(pac_u, lac);
+    }
+    //  otherwise, there's space in the scry queue; scry the lane out of ames
+    //
+    else {
+      sam_u->sat_u.foq_d++;
+      u3_noun pax = u3nq(u3i_string("peers"),
+                         u3dc("scot", 'p', u3i_chubs(2, bod_u->rec_d)),
+                         u3i_string("forward-lane"),
+                         u3_nul);
+      u3_pier_peek_last(sam_u->pir_u, u3_nul, c3__ax,
+                        u3_nul, pax, pac_u, _ames_lane_scry_cb);
     }
   }
+}
+
+/* _ames_hear(): parse a (potential packet), dispatch appropriately.
+*/
+static void
+_ames_hear(u3_ames* sam_u,
+           u3_lane* lan_u,
+           c3_w     len_w,
+           c3_y*    hun_y)
+{
+  u3_head hed_u;
+  u3_body bod_u;
+
+  //  XX packet filtering needs to revised for two protocol-change scenarios
+  //
+  //    - packets using old protocol versions from our sponsees
+  //      these must be let through, and this is a transitive condition;
+  //      they must also be forwarded where appropriate
+  //      they can be validated, as we know their semantics
+  //
+  //    - packets using newer protocol versions
+  //      these should probably be let through, or at least
+  //      trigger printfs suggesting upgrade.
+  //      they cannot be filtered, as we do not know their semantics
+  //
 
-  //  unpack the body
+  //  unpack header, ensuring buffer is large enough
   //
-  c3_y  sen_y = 2 << hed_u.sac_y;
-  c3_y  rec_y = 2 << hed_u.rac_y;
-  c3_d  sen_d[2];
-  c3_d  rec_d[2];
-  c3_w  con_w = nrd_i - 4 - sen_y - rec_y;
-  c3_y* con_y = NULL;
-
-  if ( c3y == pas_o ) {
-    u3_noun sen = u3i_bytes(sen_y, bod_y);
-    u3_noun rec = u3i_bytes(rec_y, bod_y + sen_y);
-    u3r_chubs(0, 2, rec_d, rec);
-    u3r_chubs(0, 2, sen_d, sen);
-    u3z(sen); u3z(rec);
-
-    con_y = c3_malloc(con_w);
-    memcpy(con_y, bod_y + sen_y + rec_y, con_w);
-
-    //  ensure the content is cue-able
+  if (  (4 > len_w)
+     || (c3n == _ames_sift_head(&hed_u, hun_y)) )
+  {
+    //  XX track stats
```

Comment:
Probably trivial to add, and useful statistics to have going forward.

Reply:
Added.